engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java

```diff
@@ -27,6 +29,12 @@ public final class ModuleScope implements EnsoObject {
   private Set<ModuleScope> imports;
   private Set<ModuleScope> exports;
 
+  private static final Type noTypeKey;
```

Comment:
change to
```
private static final Object NO_TYPE_KEY = new Object()
```
and use `ConcurrentHashMap<Object, ...>`. Ugly Java style, I know. But when it comes to clarity of the code I prefer encapsulation over type system.